moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -466,15 +465,31 @@ KinematicState Servo::getNextJointState(const ServoInput& command)
 
 const TwistCommand Servo::toPlanningFrame(const TwistCommand& command)
 {
+  // Reverse the twist vector since we need it to be (angular, linear) for transforming using tf2.
+  Eigen::VectorXd reordered_twist(6);
+  reordered_twist.tail<3>() = command.velocities.head<3>();
+  reordered_twist.head<3>() = command.velocities.tail<3>();
+
+  if (!servo_params_.command_frame_is_stationary)
+  {
+    RCLCPP_DEBUG_STREAM(LOGGER, "Transforming from body frame to space frame");
+    const auto body_to_space_tf =
+        transform_buffer_.lookupTransform(command.frame_id, servo_params_.ee_frame, rclcpp::Time(0));
+    tf2::doTransform(reordered_twist, reordered_twist, body_to_space_tf);
+  }
+
   if (command.frame_id != servo_params_.planning_frame)
   {
-    RCLCPP_WARN_STREAM_THROTTLE(LOGGER, *node_->get_clock(), ROS_LOG_THROTTLE_PERIOD,
-                                "Twist command is not in planning frame, transformation may not be accurate.");
+    RCLCPP_DEBUG_STREAM(LOGGER, "Transforming from " << command.frame_id << " to " << servo_params_.planning_frame);
+    const auto planning_to_command_tf =
+        transform_buffer_.lookupTransform(servo_params_.planning_frame, command.frame_id, rclcpp::Time(0));
+    tf2::doTransform(reordered_twist, reordered_twist, planning_to_command_tf);
```

Comment:
If both of these IF statements are hit, how would this be different from directly doing a single transform lookup from `servo_params_.ee_frame` to `servo_params_.planning_frame`, and should we explicitly capture this condition so we don't have to do redundant transformations + lookups on our end?

Is it true then that `if (!servo_params_.command_frame_is_stationary)`, then you're basically just always converting the command from `ee_frame` to `planning_frame` regardless or what its `command.frame_id` is? And is that correct given the spec?

Reply:
The first IF statement converts from `body` (ee frame in case of Servo) frame to `space` frame A (any stationary frame).
The second converts from the stationary frame A to planning frame.

There are 4 situations that can happen:

1. Command is with reference to planning frame and expressed in planning frame coordinates -----> None of the IF conditions are triggered
2. Command is with reference to planning frame and expressed in ee frame (body) coordinates -----> Only the first IF condition is triggered
3. Command is with reference to some stationary frame (A ) other than planning frame and expressed in the A coordinates ----->  Only the  second IF condition is triggered
4. Command is with reference to some stationary frame (A) other than planning frame and expressed in the ee frame (body) coordinates -----> both IF conditions are triggered

Reply:
Updated it.

2 & 4 are basically same,  twist expressed in body frame is independent of the space frame. so irrespective of the reference frame we can convert it to `planning_frame`.